all-sdk/external-client-configuration.md

```diff
@@ -0,0 +1,340 @@
+      implementation they have had to build in to every language.
+    * We accept that while JSON is technically human-authorable, it's not as easy as other formats. This is acceptable
+      since we will be providing tooling to mutate the config.
+    * ❓ Should we consider "json with comments" and some kind of very basic pre-parser in each language that strips
+      them? This could come later and not part of MVP.
+* This whole format will be defined in proto in the API repo in the `sdk` package.
+  * 💭 Why?
+    * Clear, centralized contract.
+    * Our SDKs already use proto dependency and this gives them stable JSON without requiring separate JSON dependency.
+    * Users can use the proto tooling to read/write configs programmatically since they are just proto JSON.
+* Do not validate that all keys are known in the file.
+  * 💭 Why?
+    * Originally this section was strict validation, but it became clear that we should allow newer-CLI-created config
+      files that may have newer keys to work with older SDKs.
+    * Similarly, it became clear that we're not validating environment variables this way, why config file keys?
+    * We want the tools that mutate this file to have strong validation on the keys being added, so hopefully this
+      avoids the concern of key typos.
+* All keys are optional in the file. Ideally in the final merged form, we'd always require `address` and we always
+  require `namespace` in namespace-specific clients (i.e. non cloud ops API), but it is unreasonable since we must fit
+  within SDK defaults.
+  * Originally this section did say `address` and `namespace` were required, but as SDK implementations came about, it
+    became clear that can't be done easily, so CLI/SDK defaults need to remain.
+* The default config file location is `~/.config/temporalio/temporal.json` for all platforms
+  * So in Windows this is often `C:\Users\MyUsername\.config\temporalio\temporal.json`
+  * ❓ Should we skip the subdir or make it `~/.config/temporal.json`? Unsure if there are any future files that would
+    go in here.
+  * Existing CLI used `~/.config/temporalio/temporal.yaml`
+    * ❓ Is this confusing for them? Are we concerned about them vs what is best moving forward?
+
+### Environment variables
+
+* Environment variables are in the form `TEMPORAL_<PROFILE_>_KEY_SUB_KEY_CAMEL_CASE`. If the profile part is missing,
+default is assumed.
+  * 💭 Why `TEMPORAL` prefix?
+    * Reasonable to differentiate.
+  * This means that `TEMPORAL_ADDRESS` and `TEMPORAL_DEFAULT_ADDRESS` are the same thing.
+  * This means that camel case parts are separated by underscores _except_ the profile name itself.
+    * So a config value of `profiles.myProfile.tls.clientCertPath` is `TEMPORAL_MYPROFILE_TLS_CLIENT_CERT_PATH`.
+    * 💭 Why?
+      * This is clearer for typing and it is unreasonable to try and break up user-supplied camel case.
+  * Also the profile may contain an underscore and retain that and match that first. So
```

Comment:
We should also limit the characters a profile name can have where we can on mutation (e.g. in the CLI) and say profiles with invalid characters in the name will fail when used in any client.

Reply:
Added outstanding question for this